spec/InterpreterSpec.hs

```diff
@@ -0,0 +1,89 @@
+{-# LANGUAGE QuasiQuotes, OverloadedStrings #-}
+
+module InterpreterSpec (spec) where
+
+import           Test.Hspec
+import           Interpreter.Mulang
+import qualified Data.Map.Strict as Map
+import           Language.Mulang.Parsers.JavaScript
+
+import           Data.Text (Text, unpack)
+import           NeatInterpolation (text)
+
+lastRef result = (\(ref, ctx) -> (Map.! ref) . globalObjects $ ctx) <$> result
+
+run expression = eval defaultContext $ js expression
+run' = run . unpack
+
+spec :: Spec
+spec = do
+  describe "evalExpr" $ do
+    it "evals addition" $ do
+      lastRef (run "1 + 2") `shouldReturn` MuNumber 3
+
+    it "evals subtraction" $ do
+      lastRef (run "2 - 1") `shouldReturn` MuNumber 1
+
+    it "evals multiplication" $ do
+      lastRef (run "2 * 3") `shouldReturn` MuNumber 6
+
+    it "evals mod" $ do
+      lastRef (run "7 % 4") `shouldReturn` MuNumber 3
+
+    it "evals and" $ do
+      lastRef (run "true && true") `shouldReturn` MuBool True
+
+    it "evals or" $ do
+      lastRef (run "false || false") `shouldReturn` MuBool False
+
+    it "evals comparison" $ do
+      lastRef (run "7 > 4") `shouldReturn` MuBool True
+
+    context "evals equal" $ do
+      it "is false when values are different" $ do
+        lastRef (run "123 == 321") `shouldReturn` MuBool False
+
+      it "is true when values are the same" $ do
+        lastRef (run "'123' == '123'") `shouldReturn` MuBool True
+
+      it "is false when values are of different types" $ do
+        lastRef (run "'123' == 123") `shouldReturn` MuBool False
+
+    context "evals if" $ do
+      it "condition is true then evals first branch" $ do
+        lastRef (run' [text|
+          if(true){
+            123
+          } else {
+            456
+          }|]) `shouldReturn` MuNumber 123
+
+      it "condition is false then evals second branch" $ do
+        lastRef (run' [text|
+          if(false){
+            123
+          } else {
+            456
+          }|]) `shouldReturn` MuNumber 456
+
+    it "condition is false then evals second branch" $ do
+      lastRef (run' [text|
+        if(false){
```

Comment:
This is Nice, but it is not how JS behaves

Reply:
Here I'm not testing against the value returned by the expression, but rather against the last reference created by the interpreter. So if the value is true, it will create a reference for the value in the first branch, and if false for the one in the second. That does not mean that that is the result of an expression.